File: src/core/ray.py
"""Ray class for raytracing"""

from .vector import Vector3


class Ray:
    """
    Represents a ray in 3D space, defined by an origin point and a direction.
    """

    def __init__(self, origin: Vector3, direction: Vector3):
        self.origin = origin
        # Normalize the direction
        self.direction = direction.normalize()

    def get_point(self, t: float) -> Vector3:
        """Get a point along the ray at parameter t"""
        return self.origin.add(self.direction.multiply(t))

File: src/core/ray.ts
import { Vector3 } from './vector';

/**
 * Represents a ray in 3D space, defined by an origin point and a direction.
 */
export class Ray {
  public direction: Vector3;

  constructor(public origin: Vector3, direction: Vector3) {
    // Normalize the direction
    this.direction = direction.normalize();
  }

  /**
   * Get a point along the ray at parameter t
   */
  getPoint(t: number): Vector3 {
    return this.origin.add(this.direction.multiply(t));
  }
}
